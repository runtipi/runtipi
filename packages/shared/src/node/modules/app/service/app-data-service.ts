import { notEmpty } from '../../../helpers/typescript-helpers';
import { DataAccessApp } from '../data-access/data-access-app';

export interface IAppDataService {
  getAppInfoFromInstalledOrAppStore: AppDataService['getAppInfoFromInstalledOrAppStore'];
  getInstalledInfo: AppDataService['getInstalledInfo'];
  getInfoFromAppStore: AppDataService['getInfoFromAppStore'];
  getUpdateInfo: AppDataService['getUpdateInfo'];
  getAllAvailableApps: AppDataService['getAllAvailableApps'];
  getAppBackups: AppDataService['getAppBackups'];
  deleteAppBackup: AppDataService['deleteAppBackup'];
}

export class AppDataService implements IAppDataService {
  private dataAccessApp: DataAccessApp;

  constructor(params: { dataDir: string; appDataDir: string; appsRepoId: string }) {
    const { dataDir, appDataDir, appsRepoId } = params;
    this.dataAccessApp = new DataAccessApp({ dataDir, appDataDir, appsRepoId });
  }

  /**
   *  This function reads the config.json and metadata/description.md files for the app with the provided id,
   *  parses the config file and returns an object with app information.
   *  It checks if the app is installed or not and looks for the config.json file in the appropriate directory.
   *  If the config.json file is invalid, it returns null.
   *  If an error occurs during the process, it logs the error message and throws an error.
   *
   *  @param {string} id - The app id.
   */
  public async getAppInfoFromInstalledOrAppStore(id: string) {
    const info = await this.dataAccessApp.getInstalledAppInfo(id);
    if (!info) {
      return this.dataAccessApp.getAppInfoFromAppStore(id);
    }
    return info;
  }

  public getInstalledInfo(id: string) {
    return this.dataAccessApp.getInstalledAppInfo(id);
  }

  public getInfoFromAppStore(id: string) {
    return this.dataAccessApp.getAppInfoFromAppStore(id);
  }

  public async getUpdateInfo(id: string) {
    return this.dataAccessApp.getAppUpdateInfo(id);
  }

  public async getAllAvailableApps() {
    const appIds = await this.dataAccessApp.getAvailableAppIds();

    const apps = await Promise.all(
      appIds.map(async (app) => {
        return this.dataAccessApp.getAppInfoFromAppStore(app);
      }),
    );

    return apps.filter(notEmpty).map(({ id, categories, name, short_desc, deprecated, supported_architectures, created_at }) => ({
      id,
      categories,
      name,
      short_desc,
      deprecated,
      supported_architectures,
      created_at,
    }));
  }

  public async getAppBackups(params: { appId: string; pageSize: number; page: number }) {
    const { appId, page, pageSize } = params;
    const backups = await this.dataAccessApp.listBackupsByAppId(appId);
    backups.sort((a, b) => b.date.getTime() - a.date.getTime());

    const start = (page - 1) * pageSize;
    const end = start + pageSize;
    const data = backups.slice(start, end);

    return {
      data,
      total: backups.length,
      currentPage: Math.floor(start / pageSize) + 1,
      lastPage: Math.ceil(backups.length / pageSize),
    };
  }

  public async deleteAppBackup(appId: string, filename: string) {
    return this.dataAccessApp.deleteBackup(appId, filename);
  }
}
