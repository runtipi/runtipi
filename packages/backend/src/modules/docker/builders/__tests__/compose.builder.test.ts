import { beforeEach, describe, expect, it } from 'vitest';
import { DockerComposeBuilder } from '../compose.builder';
import type { ServiceInput } from '../schemas';
import { ServiceBuilder } from '../service.builder';

describe('DockerComposeBuilder', () => {
  let composeBuilder: DockerComposeBuilder;
  let serviceBuilder: ServiceBuilder;

  beforeEach(() => {
    composeBuilder = new DockerComposeBuilder();
    serviceBuilder = new ServiceBuilder();
  });

  it('should build a docker-compose file', () => {
    const serviceName = 'service';
    const service: ServiceInput = {
      name: serviceName,
      image: 'image',
    };

    const compose = composeBuilder.getDockerCompose([service], {}, 'test');
    expect(compose).toMatchSnapshot();
  });

  it('should correctly format deploy resources', () => {
    const service: ServiceInput = {
      name: 'service',
      image: 'image',
      deploy: {
        resources: {
          limits: { cpus: '0.50', memory: '50M', pids: 1 },
          reservations: { cpus: '0.25', memory: '20M', devices: [{ capabilities: ['gpu'], driver: 'nvidia', count: 'all' }] },
        },
      },
    };

    const compose = composeBuilder.getDockerCompose([service], {}, 'test');
    expect(compose).toMatchSnapshot();
  });

  it('should correctly format devices', () => {
    const service: ServiceInput = {
      name: 'service',
      image: 'image',
      devices: ['/dev/ttyUSB0:/dev/ttyUSB0', '/dev/sda:/dev/xvda:rwm'],
    };

    const compose = composeBuilder.getDockerCompose([service], {}, 'test');
    expect(compose).toMatchSnapshot();
  });

  it('should correctly format entrypoint as string', () => {
    const service: ServiceInput = {
      name: 'service',
      image: 'image',
      entrypoint: 'entrypoint',
    };

    const compose = composeBuilder.getDockerCompose([service], {}, 'test');

    expect(compose).toMatchSnapshot();
  });

  it('should correctly format entrypoint as array', () => {
    const service: ServiceInput = {
      name: 'service',
      image: 'image',
      entrypoint: ['entrypoint', 'arg1', 'arg2'],
    };

    const compose = composeBuilder.getDockerCompose([service], {}, 'test');

    expect(compose).toMatchSnapshot();
  });

  it('should correctly format logging', () => {
    const service: ServiceInput = {
      name: 'service',
      image: 'image',
      logging: { driver: 'json-file', options: { 'syslog-address': 'tcp://192.168.0.42:123' } },
    };

    const compose = composeBuilder.getDockerCompose([service], {}, 'test');

    expect(compose).toMatchSnapshot();
  });

  it('should correctly format a complex docker-compose file', () => {
    const service1: ServiceInput = {
      name: 'service1',
      image: 'image1',
      internalPort: 80,
      addPorts: [{ containerPort: 8080, hostPort: 3400 }],
      extraHosts: ['host1', 'host2'],
      ulimits: { nproc: 1024, nofile: 65536 },
      command: 'node index.js',
      volumes: [
        { hostPath: '/host/path', containerPath: '/container/path', readOnly: true },
        { hostPath: '/host/path2', containerPath: '/container/path2' },
      ],
      environment: { NODE_ENV: 'production', PORT: 80, SOME_VAR: 'value' },
      healthCheck: { test: 'curl -f http://localhost/ || exit 1', interval: '1m30s', timeout: '10s', retries: 3, startPeriod: '40s' },
      dependsOn: ['service2'],
      capAdd: ['SYS_ADMIN', 'NET_ADMIN'],
      deploy: {
        resources: {
          limits: { cpus: '0.50', memory: '50M', pids: 1 },
          reservations: { cpus: '0.25', memory: '20M', devices: [{ capabilities: ['gpu'], driver: 'nvidia', count: 'all' }] },
        },
      },
      hostname: 'hostname',
      devices: ['/dev/ttyUSB0:/dev/ttyUSB0', '/dev/sda:/dev/xvda:rwm'],
      entrypoint: ['entrypoint', 'arg1', 'arg2'],
      pid: '1',
      privileged: true,
      tty: true,
      user: 'user',
      workingDir: '/working/dir',
      shmSize: '1G',
      capDrop: ['SYS_ADMIN', 'NET_ADMIN'],
      logging: { driver: 'json-file', options: { 'syslog-address': 'tcp://192.168.0.42:123' } },
      readOnly: true,
      securityOpt: ['label=disable', 'label=role:ROLE'],
      stopSignal: 'SIGTERM',
      stopGracePeriod: '1m',
      stdinOpen: true,
    };

    const service2: ServiceInput = {
      name: 'service2',
      image: 'image2',
    };

    const compose = composeBuilder.getDockerCompose([service1, service2], {}, 'test');

    expect(compose).toMatchSnapshot();
  });

  it('should add correct traefik labels to the main service', () => {
    const service: ServiceInput = {
      name: 'service',
      image: 'image',
      internalPort: 440,
      isMain: true,
    };

    const compose = composeBuilder.getDockerCompose([service], { exposed: true, exposedLocal: true, openPort: true }, 'test');

    expect(compose).toMatchSnapshot();
  });
});
