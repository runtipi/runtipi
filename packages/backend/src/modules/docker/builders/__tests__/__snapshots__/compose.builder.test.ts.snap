// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DockerComposeBuilder > should add correct traefik labels to the main service 1`] = `
"services:
  service_test:
    image: image
    container_name: service_test
    restart: unless-stopped
    networks:
      - tipi_main_network
    ports:
      - \${APP_PORT}:440
    labels:
      generated: true
      traefik.enable: true
      traefik.http.middlewares.service_test-web-redirect.redirectscheme.scheme: https
      traefik.http.services.service_test.loadbalancer.server.port: "440"
      traefik.http.routers.service_test-insecure.rule: Host(\`\${APP_DOMAIN}\`)
      traefik.http.routers.service_test-insecure.entrypoints: web
      traefik.http.routers.service_test-insecure.service: service_test
      traefik.http.routers.service_test-insecure.middlewares: service_test-web-redirect
      traefik.http.routers.service_test.rule: Host(\`\${APP_DOMAIN}\`)
      traefik.http.routers.service_test.entrypoints: websecure
      traefik.http.routers.service_test.service: service_test
      traefik.http.routers.service_test.tls.certresolver: myresolver
      traefik.http.routers.service_test-local-insecure.rule: Host(\`service-test.\${LOCAL_DOMAIN}\`)
      traefik.http.routers.service_test-local-insecure.entrypoints: web
      traefik.http.routers.service_test-local-insecure.service: service_test
      traefik.http.routers.service_test-local-insecure.middlewares: service_test-web-redirect
      traefik.http.routers.service_test-local.rule: Host(\`service-test.\${LOCAL_DOMAIN}\`)
      traefik.http.routers.service_test-local.entrypoints: websecure
      traefik.http.routers.service_test-local.service: service_test
      traefik.http.routers.service_test-local.tls: true
networks:
  tipi_main_network:
    name: runtipi_tipi_main_network
    external: true
"
`;

exports[`DockerComposeBuilder > should build a docker-compose file 1`] = `
"services:
  service_test:
    image: image
    container_name: service_test
    restart: unless-stopped
    networks:
      - tipi_main_network
networks:
  tipi_main_network:
    name: runtipi_tipi_main_network
    external: true
"
`;

exports[`DockerComposeBuilder > should correctly format a complex docker-compose file 1`] = `
"services:
  service1_test:
    image: image1
    command: node index.js
    container_name: service1_test
    restart: unless-stopped
    networks:
      - tipi_main_network
    extra_hosts:
      - host1
      - host2
    ulimits:
      nproc: 1024
      nofile: 65536
    healthcheck:
      test: curl -f http://localhost/ || exit 1
      interval: 1m30s
      timeout: 10s
      retries: 3
      start_period: 40s
    environment:
      NODE_ENV: production
      PORT: 80
      SOME_VAR: value
    ports:
      - 3400:8080
    volumes:
      - /host/path:/container/path:ro
      - /host/path2:/container/path2
    depends_on:
      - service2
    cap_add:
      - SYS_ADMIN
      - NET_ADMIN
    deploy:
      resources:
        limits:
          cpus: "0.50"
          memory: 50M
          pids: 1
        reservations:
          cpus: "0.25"
          memory: 20M
          devices:
            - capabilities:
                - gpu
              driver: nvidia
              count: all
    hostname: hostname
    devices:
      - /dev/ttyUSB0:/dev/ttyUSB0
      - /dev/sda:/dev/xvda:rwm
    entrypoint:
      - entrypoint
      - arg1
      - arg2
    pid: "1"
    privileged: true
    tty: true
    user: user
    working_dir: /working/dir
    shm_size: 1G
    cap_drop:
      - SYS_ADMIN
      - NET_ADMIN
    logging:
      driver: json-file
      options:
        syslog-address: tcp://192.168.0.42:123
    read_only: true
    security_opt:
      - label=disable
      - label=role:ROLE
    stop_signal: SIGTERM
    stop_grace_period: 1m
    stdin_open: true
  service2_test:
    image: image2
    container_name: service2_test
    restart: unless-stopped
    networks:
      - tipi_main_network
networks:
  tipi_main_network:
    name: runtipi_tipi_main_network
    external: true
"
`;

exports[`DockerComposeBuilder > should correctly format deploy resources 1`] = `
"services:
  service_test:
    image: image
    container_name: service_test
    restart: unless-stopped
    networks:
      - tipi_main_network
    deploy:
      resources:
        limits:
          cpus: "0.50"
          memory: 50M
          pids: 1
        reservations:
          cpus: "0.25"
          memory: 20M
          devices:
            - capabilities:
                - gpu
              driver: nvidia
              count: all
networks:
  tipi_main_network:
    name: runtipi_tipi_main_network
    external: true
"
`;

exports[`DockerComposeBuilder > should correctly format devices 1`] = `
"services:
  service_test:
    image: image
    container_name: service_test
    restart: unless-stopped
    networks:
      - tipi_main_network
    devices:
      - /dev/ttyUSB0:/dev/ttyUSB0
      - /dev/sda:/dev/xvda:rwm
networks:
  tipi_main_network:
    name: runtipi_tipi_main_network
    external: true
"
`;

exports[`DockerComposeBuilder > should correctly format entrypoint as array 1`] = `
"services:
  service_test:
    image: image
    container_name: service_test
    restart: unless-stopped
    networks:
      - tipi_main_network
    entrypoint:
      - entrypoint
      - arg1
      - arg2
networks:
  tipi_main_network:
    name: runtipi_tipi_main_network
    external: true
"
`;

exports[`DockerComposeBuilder > should correctly format entrypoint as string 1`] = `
"services:
  service_test:
    image: image
    container_name: service_test
    restart: unless-stopped
    networks:
      - tipi_main_network
    entrypoint: entrypoint
networks:
  tipi_main_network:
    name: runtipi_tipi_main_network
    external: true
"
`;

exports[`DockerComposeBuilder > should correctly format logging 1`] = `
"services:
  service_test:
    image: image
    container_name: service_test
    restart: unless-stopped
    networks:
      - tipi_main_network
    logging:
      driver: json-file
      options:
        syslog-address: tcp://192.168.0.42:123
networks:
  tipi_main_network:
    name: runtipi_tipi_main_network
    external: true
"
`;
